### PR TITLE
Unified Memory Manager design proposal

### PR DESCRIPTION
**Pull request by philipportner**
Here is the design proposal for the Unified Memory Managed we discussed.
I would greatly appreciate any feedback on stuff that I may have misunderstood.
Also please point out any information this document should contain but is missing.

### CODE REVIEW

File: scripts/staging/unified-memory-manager/umm.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+### Block Java Class
+The first implementation will only handle CacheBlock objects, at the moment
+without any major changes to the CacheBlock object itself.
+
+If there is enough time this will be extended to an additional block type, `OperandBlock`.
+This would either require a new data type or maybe we could reuse some of the ideas
+from the `CacheableData<T extends CacheBlock> extends Data` class.
+
+### Testing
+Testing will be done with the existing java based test system which runs a
```
Comment (mboehm7) [inline scripts/staging/unified-memory-manager/umm.md:124]: For this exploratory project, I would recommend to only implement the memory manager, and add component tests that uses certain sequences of interactions and checks the final state. That way you can instantiate the memory manager with a small memory budget and test everything with tiny objects.

**Review by mboehm7 — COMMENTED** [review]
Thanks @philipportner - that's a good starting point. Please incorporate the feedback and then I'll merge it in.

File: scripts/staging/unified-memory-manager/umm.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ * Loads the block into memory and pins it, if not evicted just pins it(?)
+ * The requested block needs to fit into UMM boundaries.
+ */
+Block serve(String);
+
+/**
+ * Reserves an specified amount in one of the two memory areas.
+ * The requested block needs to fit into UMM boundaries.
+ */
+Block reserve(Size, MemoryArea);
```
[the remark this answers is not in the document]
Comment (philipportner) [reply scripts/staging/unified-memory-manager/umm.md:81]: Yes, I thought one may want the handle as soon as the memory is reserved.  If that's not the case I'll make it `void reserve(Size)`.

File: scripts/staging/unified-memory-manager/umm.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * Pins the input block into main memory.
+ * The provided block needs to fit into UMM boundaries.
+ */
+void pin(String, Block);
+
+/**
+ * Unpins the input block from main memory.
+ */
+void unpin(String);
```
[the remark this answers is not in the document]
Comment (philipportner) [reply scripts/staging/unified-memory-manager/umm.md:69]: Not quite sure how I would merge that with `resize_reserved`.
`void unpin(String, Size)` and move the block from operation memory to the buffer pool and drop/resize if already in buffer pool?

File: scripts/staging/unified-memory-manager/umm.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+occupy, meaning that the boundary for the areas can shift dynamically depending
+on the current load.
+
+||min|max|
+| ------------- |:-------------:| -----:|
+| operations  | 50% | 70% |
+| buffer pool | 15% | 30% |
+
+The UMM will utilise the existing `CacheBlock` structure to manage the buffer
+pool area while it will use the new `OperandBlock (?)` structure to keep track of
```
[the remark this answers is not in the document]
Comment (philipportner) [reply scripts/staging/unified-memory-manager/umm.md:53]: I may be using the terminology wrong, I actually meant what you described.

I looked through the code again and I think I'm missing something.

The `MatrixObject`, `FrameObject` and the `TensorObject` implement the `CacheBlock` interface.
The `LazyWriteBuffer` itself seems to manage `ByteBuffer` instances, which in turn also hold `CacheBlocks`, at least for dense data.

Could you please elaborate again on a few things I seem to not be 100% clear about.
1) Which current classes do you mean when you talk about operation memory?
2) Which classes are we talking about when you mean intermediates / dirty objects that are supposed to go into the buffer pool area?

I'll try to answer them myself, maybe you can correct me.
ad1) As mentioned in the design document, derivatives of `Data` like `MatrixObject`, `FrameObject` and `TensorObject`.
ad2) Here I'm not sure, looking at the code I would say these are also `CacheableData` objects, for example in the `ExecutionContext` or `EstimatorDensityMap`, these are working with `MatrixBlocks`, which ultimately is also a `CacheBlock`. Only dense data, e.g., `DenseBlock` instances do not hold `CacheBlock`.

It kinda feels like I'm missing something, we said the second `Block` does not exist currently but it looks like the `CacheBlock` would suffice. In that case the split memory area would just be a logical differentiation.

**Comment by philipportner** [comment]
> Thanks @philipportner - that's a good starting point. Please incorporate the feedback and then I'll merge it in.

Thanks for the feedback @mboehm7 .
Seems like I still have some questions and added them to your comments.
I would appreciate if you could answer them.

**Comment by mboehm7** [comment]
LGTM - thanks @philipportner. Following our offline discussion, I think it's best to start bottom-up with this very central component. So for now we focus purely on the cache blocks and implement and test the memory manager in a component-local manner without considering the surrounding meta data objects (CacheableData). 

I now converted you design document to a java class with the design as java doc and a reworked API that should cover the main use cases and clarify the individual API interactions and state transitions. Looking forward to your initial implementation.

**Pull request closed without merging**